File: 547-number-of-provinces/547.py
#!/usr/bin/env python
# -*- coding:utf-8 -*-

from typing import List
from collections import deque

class Solution:
    def findCircleNum(self, isConnected: List[List[int]]) -> int:
        size = len(isConnected)
        arr = [True] * size
        dq = deque()
        res = 0
        for i in range(size):
            if arr[i]:
                arr[i] = False
                dq.append(i)
                while dq:
                    left = dq.popleft()
                    for j in range(0, size):
                        if isConnected[left][j] and arr[j]:
                            arr[j] = False
                            dq.append(j)
                res += 1
        return res

if __name__ == '__main__':
    sol = Solution()
    isConnected = [[1,1,0],[1,1,0],[0,0,1]]
    print(sol.findCircleNum(isConnected))
    isConnected = [[1,0,0],[0,1,0],[0,0,1]]
    print(sol.findCircleNum(isConnected))

File: 547-number-of-provinces/547.cpp
#include <iostream>
#include <vector>
#include <queue>

using namespace std;

class Solution {
public:
    int findCircleNum(vector<vector<int>>& isConnected) {
        int len = isConnected.size();
        if (len == 0) return 0;
        int res = 0;
        vector<bool> vec = vector<bool>(len, true);
        queue<int> q;
        for (int i = 0; i < len; ++i) {
            if (vec[i]) {
                vec[i] = false;
                q.push(i);
                while (!q.empty()) {
                    int pop = q.front();
                    q.pop();
                    for (int j = 0; j < len; ++j) {
                        if (isConnected[pop][j] == 1 && vec[j]) {
                            vec[j] = false;
                            q.push(j);
                        }
                    }
                }
                ++res;
            }
        }
        return res;
    }
};

int main(int argc, char** args) {
    Solution sol = Solution();
    vector<vector<int>> isConnected1 = {{1,1,0},{1,1,0},{0,0,1}};
    cout << sol.findCircleNum(isConnected1) << endl;
    vector<vector<int>> isConnected2 = {{1,0,0},{0,1,0},{0,0,1}};
    cout << sol.findCircleNum(isConnected2) << endl;
    return 0;
}